### ISSUE TITLE
Fixing Sound Glitch by Padding Numpy Array for Sound Play 

### ISSUE THREAD

**Issue by waszee** [spatialaudio/python-sounddevice#282]
I want to transition a morse code generator that  works using pygame sndarray to using sounddevice play.  However sd.play is clipping or glitching the end of the letter sounds often making dahs sound like dits.  Based on earlier comments about glitches I am pretty sure the problem is a mismatch between the length of the Numpy array and the buffer size for Sd.play.  Sd.play did not seem to have a buffer clear command that I could use so sometimes the buffer appears to be carried over to next letter.  The code uses a dictionary to convert letters to the the morse code dit and dah pattern.  That pattern as shown below is built into a letter numpy array.   This is done by merging numpy arrays containing sine wave data of the appropriate length for the dit, dah, and spaces for a desired letter,  given a tone frequency and sound card sampling rate. There are many examples for doing this on github.   In pygame the numpy letter was sent to pygame.sndarray.make.sound  before playing and that worked but sending the letter arrary directly to sd.play{} is not reliable.

I think I need to pad or vary the numpy array size to match what sd.play wants.  I can do this by padding the intraspace delays but I am not sure how to calculate and hope one of you experts can advise.

[sd_play_morse_glitches.zip](https://github.com/spatialaudio/python-sounddevice/files/5251058/sd_play_morse_glitches.zip

So how do I convert a Numpy arrary to a perfect match for sd.play?

**Comment by mgeier** [comment]
Can you please paste the contents of the linked file as a comment (with proper formatting)?
See also https://python-sounddevice.readthedocs.io/en/0.4.0/CONTRIBUTING.html#reporting-problems.

Not having looked at the code, it might also be interesting to see your list of devices and it might be interesting to try different host APIs to see whether the glitches appear in all of them.

**Comment by waszee** [comment]
The following is playing wrong, more so on the Windows machine than the Raspberry Pi.  I am developing code on both a raspberry pi 4 and a win10 computer.  The latter is running anaconda and spyder.  The device list on the Windows machine looks like this.  I sometimes add a USB sound device which is required to add a microphone input on the Pi.  Output is generally with the HDMI connected Samsung monitor but the clip problem occurs with the other output options.  The word "Hello" or "50" is good for a test.

   0 Microsoft Sound Mapper - Output, MME (0 in, 2 out)
<  1 1 - SAMSUNG (AMD High Definitio, MME (0 in, 2 out)
   2 Digital Audio (S/PDIF) (High De, MME (0 in, 2 out)
   3 Headphones (High Definition Aud, MME (0 in, 2 out)
   4 Primary Sound Driver, Windows DirectSound (0 in, 2 out)
   5 1 - SAMSUNG (AMD High Definition Audio Device), Windows DirectSound (0 in, 2 out)
   6 Digital Audio (S/PDIF) (High Definition Audio Device), Windows DirectSound (0 in, 2 out)
   7 Headphones (High Definition Audio Device), Windows DirectSound (0 in, 2 out)
   8 Digital Audio (S/PDIF) (High Definition Audio Device), Windows WASAPI (0 in, 2 out)
   9 1 - SAMSUNG (AMD High Definition Audio Device), Windows WASAPI (0 in, 2 out)
  10 Headphones (High Definition Audio Device), Windows WASAPI (0 in, 2 out)
  11 SPDIF Out (HD Audio SPDIF out), Windows WDM-KS (0 in, 2 out)
  12 Microphone (HD Audio Microphone), Windows WDM-KS (2 in, 0 out)
  13 Headphones (HD Audio Headphone), Windows WDM-KS (0 in, 2 out)
  14 Output (AMD HD Audio HDMI out #0), Windows WDM-KS (0 in, 2 out)


# -*- coding: utf-8 -*- 
"""
Created on Tue Sep 15 18:31:57 2020
The Coded is GNU license 3 with no warranty implied. Use at own risk.
@author: waszee
"""

import sounddevice as sd
import sys
import numpy as np 
import time
#from time import sleep
#import matplotlib.pyplot as plt


CODE = {'A': '.-',     'B': '-...',   'C': '-.-.', 
        'D': '-..',    'E': '.',      'F': '..-.',
        'G': '--.',    'H': '....',   'I': '..',
        'J': '.---',   'K': '-.-',    'L': '.-..',
        'M': '--',     'N': '-.',     'O': '---',
        'P': '.--.',   'Q': '--.-',   'R': '.-.',
        'S': '...',    'T': '-',      'U': '..-',
        'V': '...-',   'W': '.--',    'X': '-..-',
        'Y': '-.--',   'Z': '--..',
        
        '0': '-----',  '1': '.----',  '2': '..---',
        '3': '...--',  '4': '....-',  '5': '.....',
        '6': '-....',  '7': '--...',  '8': '---..',
        '9': '----.',
        
        ',':'--..--',  '.':'.-.-.-',  '?':'..--..',
        '/':'-..-.',   '-':'-....-',  '!':'-.-.--',
        '(':'-.--.',   ')':'-.--.-',  '&':'.-....',
        ':':'---...',  ';':'-.-.-.',  '=':'-...-',
        '+':'.-.-.',   '_':'..--.-',  '"':'.-..-.',
        '$':'...-..-', '@':'.--.-.',
        
        '<':'.-...',  '>':'.-.-.',
        '%':'...-.-', 
        '{':'....--',  '}':'...-.',
        
        
        "'":'.----.',
        
        '\n':'-...-.-', #line feed = <BK>  same as \x0a
        
        '\u00c4':'.-.-', #a umlat
        '\u00d6':'---.', #o umlat
        '\u00dc':'..--', #u umlat
        '\u03b2':'...--..', # beta 
        }
  
 
def verify(string):
    keys = CODE.keys()
    for char in string:
        if char.upper() not in keys and char != ' ':
            sys.exit('Error - a character ' + char + ' in the msg cannot be translated to Morse Code')


def beat(tlen,samrate,tone):
    t = np.arange(0,tlen,1/samrate) 
    asnd = 0.4 * np.sin(2*np.pi*tone*t)   # 0.5 is arbitrary to avoid clipping sound card DAC
    asnd  = (asnd*32768).astype(np.int16)
    return asnd                     

def main():
        
    samrate = 44100 # audio sampling rate
    T = 0.06 # .08=10wpm, 0.07=14wpm. 0.06=17wpm, 0.05=20 wpm, 
    T3 = 3*T #
    #farn=1.0 #  a modifier like Farnsworth  to lengthen space between sounds or letters
    tone = 750 # audio tone freq
    sd.default.samplerate = samrate
    sd.default.channels = 1
    sd.default.dtype = np.int16
    
    #pygame.mixer.pre_init(samrate, size=-16, channels=1)
    #pygame.mixer.init()
    #pygame.init()    
   
   #createHi  dit, day and intraspace as numpy array values
    dit = beat(T,samrate,tone)
    dah = beat(T*3,samrate,tone)
    space = np.zeros(dit.size,np.int16) 
    
    print ('Welcome to Python Code Out ver 0.2\n')
    print('Types exit to exit message loop or ^C to abort')
    try:  
       
        while True:
                     
            msg = input('Enter Message: ')
             
            verify(msg)
            print ()
           
            if msg == 'exit': break      
            for char in msg:
                if char == ' ':
                        print (' '*7)
                        time.sleep(T*7) #word space delay
                else:
                    print (char + ' ' + CODE[char.upper()])
                    c=space
                    for char in CODE[char.upper()]:
                       
                        if char == '.':
                            b=dit
                        else:
                            b=dah
                        c = np.concatenate((c,space,b))
                     
                    sd.play(c, blocking = True)
                                   
                time.sleep(T3)                       
                    #pygame.time.wait(int(T3*1000)) #ketter spacing interval
                     
  
    except KeyboardInterrupt:
        print('interupt by keyboard control c detected')
        
    except Exception as e:
        print('Exception: ' and e)
            
    print ('\n\nGoodbye!')
  
                   
        
if __name__ == "__main__":
    main()

**Comment by waszee** [comment]
Matthias,

Hope you can take a look at the code.  I updated the post.  The direct cut and paste resulted in some form of auto clustering but I think should be easy to correct if someone wants to download.  I can send more details including sound capture or the pygame version if needed but was trying to keep post reasonable.

Thanks,

Bill Spencer, [posted as waszee]
Aiken, SC

**Comment by mgeier** [comment]
> I updated the post.

Thanks, it is still not quite readable.

> The direct cut and paste resulted in some form of auto clustering but I think should be easy to correct if someone wants to download.

You should correct that.

Can you try to insert ` ``` ` in a separate line before and after each code (and terminal output) block?

An example is shown in https://python-sounddevice.readthedocs.io/en/0.4.0/CONTRIBUTING.html#reporting-problems.

> I can send more details including sound capture or the pygame version if needed but was trying to keep post reasonable.

If you think there are too many details, you can hide them with the `<details>` tag, which is also explained in https://python-sounddevice.readthedocs.io/en/0.4.0/CONTRIBUTING.html#reporting-problems.

---

I still didn't really look at your code, but I've seen that you use `sd.play()`. As mentioned in the docs, this is a simple convenience function which might not be suitable for more complicated use cases.
I think your example is such a case where it is not suitable.

Apart from the glitches, I guess the timing of the dits and dahs is also not great, right?

It would be better to only use one "stream" object which is started in the beginning, kept open for the whole transmission and then closed in the end.

I normally like to implement such things with a callback function, but this Morse code application could probably be implemented more easily using the so-called "blocking" API. This should allow you to keep the stream active and at the same time get exact timing between the dits and dahs.

**Comment by waszee** [comment]
Matthias,

Thank you for the advice. I will look into the blocking API and have been studying your callback methods.

I have another need for your advice.  I am not sure how to post or if should.  I have been trying to run your “plotinput.py” script, which you have posted as an example.  The program runs fine on my raspberry Pi 4 except I had to change the download interval from 30 to about 50 to prevent overloading and plot spikes but program ran as expected. However, on my Windows 10 system with Anaconda and Spyder(Python 3.8) I am having serious trouble.  I actually got program to run twice and partially with a quick freeze up using a stepwise debug but most times nothing is captured.  There appears to be two problems which might be related but I don’t think so.

First, with the arg.channels error check statement, I get a startup failure that c must be >=1.   However the windows version seems to be messing up the parser arg somehow.  All of the other values  are as expected but the channels args are messing up.  Actually I see nothing wrong in your program….

[cid:image001.png@01D6940C.FA4A73B0]

The second problem occurs if I simply comment out the check channels query as I know the default is okay.  I let the program run

#if any(c < 1 for c in args.channels):
#    parser.error('argument CHANNEL: must be >= 1')

I then get the Figure 1 to pop-up and no error messages, but the figure does not show captured data and does not roll or update.  Program just stops.
[cid:image002.png@01D6940D.983A4690]

If I run in debug mode and put a halt at the plt.show() statement then some data is captured but if I drop the debug halt the program again fails.  I have tried to change the queue.Empty exception to wait till something in the queue, but that did not work for me.  Please let me know if you have suggestions.  The program runs fine on the Raspian 64 linux that I run.  Your record to file demo program also works in Windoews.

[cid:image003.png@01D6940E.8A51D4C0]


Thanks for the support,
Bill, waszee,

**Comment by waszee** [comment]
Matthias - thank you for the comments.  I hope this feedback helps makes your module better.  I hope to use especially for input capturing - thanks for the product...

FYI - I modified my script to ramp the sound pulses up and down for dits and dahs and that greatly improved the output sound.  I also modified the code to round the numpy arrays to the nearest multiple of 1024 for each letter.  That did not seem to help.  I posted a version of the code using the pygame module on Github  search waszee/SignalStudies_RF repository.  After the improvements SndPlay handled individual code letters fine but code sequence such as "gfd"  or "abc" resulted in some of the letter dits being skipped or a dah possibly being clipped to a dit.  This occurred sometimes in the middle of the sound.  I suspect I might get this to work using the callback approach as you suggest  but will need to study that more.  I have made enough progress to move on. to other aspects of my study.  I will drop out for now.

**Comment by waszee** [comment]
Matthias,

Quick  question – I suspect this is an obvious issue to you but it tripped a newbie like me.

I was cutting and pasting from your program examples for Sounddevice and encountered an issue. I get a  PortAudio error with the inputstream  “channel” specification. It occurs on both a raspberry pi and in Anaconda Spyder on a Windows machine.

I found I needed to use:

InputStream(…,  channels = args.channels, ….) where args.channels  =1 or defaulted to 1 in the arguments.  It works.

But,  if I tried to set channels directly, that is :

inputStream(….., channels = 1, …..) the program crashes.  On the raspberry I get a PortAudio error.  Program freezes on Anaconda.

All of the other stream variables seem to allow a direct integer as input, for example sample_rate=44100 is okay. What is different about channels?  Is this a numpy dimension thing or something about what is expected in your call.


Bill Spencer
Aiken, SC

**Comment by mgeier** [comment]
There should be nothing special about `channels`, the problem is nearly certainly something else.

Please provide a minimal reproducible example, ideally in a new issue, see also https://python-sounddevice.readthedocs.io/en/0.4.1/CONTRIBUTING.html#reporting-problems.

**Comment by waszee** [comment]
Thanks for quick response.  I did fix some other stuff and can no longer reproduce on my Pi.  Will post if see again.

FYI  I loaded a script that uses sounddevice to display an audio spectrogram on waszee SignalStudies_RF.  The program is cw_spectrogram_v2_snddev.  It plots input audio signals [in my case morse code from a radio] providing realtime display of the input freq domain as well as secondary plot of the spectrogram over 30 seconds.  I did have to play with the blocksize some but basically followed some previously posted code using pyaudio.  Still a work in progress but starting to work.  Used a class structure this time which might be of interest to some other newbies.

Bill

**Comment by mgeier** [comment]
Here a clickable link: https://github.com/waszee/SignalStudies_RF

**Comment by waszee** [comment]
Thanks I was wondering about how I should do that.   Will use the link reference.

FYI I decided to try the script on a Ubuntu laptop system this morning.  I installed PyCharm CE as the IDE.  The debugger is much better than the Thonny I was using on the raspberry pi and I was able to figure out how I likely messed up the channel number in my previous question to you.  I had changed a datatype that was expected to be a str to a short [int16] and everything was shifting two bytes on me.

I decided to load PyCharm on my windows machine and some problems I had with Anaconda Spyder seem less of a problem and version control and formatting should get better.  Working on trying to make sliding images inside a window and scaling versus the FFT window and the sampling rate.  It is a somewhat tricky balance but so far your callback and input code is working for me.

Bill [or waszee or KA4ZBN]
